### PR TITLE
Block Editor: Optimize layout style renderer subscription

### PR DESCRIPTION
**Pull request by Mamaduka**
## What?
This is a follow-up to https://github.com/WordPress/gutenberg/pull/55754#pullrequestreview-1707861987.
Similar to #55449.

PR moves most of the logic from `withLayoutStyles` and `withChildLayoutStyles` into separate components, which are only rendered when a block supports layout.

Tested using @jsnajdr's debug code (https://github.com/WordPress/gutenberg/commit/c02930390ee76935cc545a708eefe6d9b120f240), the store subscriptions are reduced by 4000 on large text post - 1000 blocks, i.e., four subscription per-block.

## Why?
See https://github.com/WordPress/gutenberg/pull/54819#issuecomment-1761202859.

## Testing Instructions
1. Open a post or page.
2. Insert Row or Stack blocks, add child block, and test various layout settings.
3. Confirm they work as before.

### Testing Instructions for Keyboard
Same.

### CODE REVIEW

**Comment by Mamaduka** [comment]
@andrewserong, @tellthemachines, I tried to be thorough when testing the changes. Do you have patterns you use for testing layout changes? Those would be helpful here.

File: packages/block-editor/src/hooks/layout.js
Comment (jsnajdr) [inline packages/block-editor/src/hooks/layout.js]: The name should be either `BlockWith...` or `BlockListBlockWith...`. The thing is a "block", not a "list".

File: packages/block-editor/src/hooks/layout.js
Comment (jsnajdr) [inline packages/block-editor/src/hooks/layout.js]: Pre-existing code style issue: `const layout = props.attributes.style?.layout ?? {}` is easier to read IMO.

File: packages/block-editor/src/hooks/layout.js
Comment (jsnajdr) [inline packages/block-editor/src/hooks/layout.js]: Both assignments to the `css` variable can use `css =` instead of `css +=`, the previous value of `css` is always empty.

File: packages/block-editor/src/hooks/layout.js
Comment (jsnajdr) [inline packages/block-editor/src/hooks/layout.js]: `props` are always an object, can be used as `props.className`, without the `?.` operator.

File: packages/block-editor/src/hooks/layout.js
Comment (jsnajdr) [inline packages/block-editor/src/hooks/layout.js:474]: There's one thing about the child layout code that is buggy, or at least inefficient: if `disableLayoutStyles` is true, and if the block has a child layout (i.e., there are valid `flexSize` or `selfStretch` values), the corresponding CSS will be created and added to style overrides. But the element _won't_ get the `wp-container-content-n` CSS class, and therefore the generated CSS won't be applied to any element, it's redundant.

The `withLayoutStyles` HOC is more efficient: it's doesn't even generate the CSS if layout styles are disabled. It would be nice if `withChildLayoutStyles` behaved the same way.

Cc @tellthemachines to verify that the argument above is correct and that there's no hidden gotcha.

File: packages/block-editor/src/hooks/layout.js
Comment (jsnajdr) [inline packages/block-editor/src/hooks/layout.js]: The `disableLayoutStyles` could be efficiently moved into the parent HOC component, leading to rendering the `BlockWithLayoutStyles` component less often. Just use this pattern:
```js
const hasChildLayout = ...;
const shouldRenderLayoutStyles = useSelect( ( select ) => {
  if ( ! hasChildLayout ) {
    return false; // return early before calling `select` at all
  }
  return ! select( blockEditorStore ).getSettings().disableLayoutStyles;
}, [ hasChildLayout ] );
```
The trick is that while `hasChildLayout` remains `false`, the `select` function is not called and no store subscriptions are created.

**Review by jsnajdr — APPROVED** [review]
This is OK and possible to merge, but I noticed quite a few non-blocking pre-existing issues. Some are just code style nits, others are more impactful.

As a followup, would it make sense to merge the `withLayoutStyles` and `withChildLayoutStyles` HOCs? Is there some common code that they both need to execute, and that could be executed just once instead of twice in the merged component?

File: packages/block-editor/src/hooks/layout.js
Comment (Mamaduka) [reply packages/block-editor/src/hooks/layout.js]: Great suggestion! I forgot about this pattern, even though I read your previous PRs 😓

**Comment by Mamaduka** [comment]
Thanks for the great feedback, @jsnajdr!

I've addressed most of it, besides the https://github.com/WordPress/gutenberg/pull/55762#discussion_r1379697523. I'll wait for additional details from @tellthemachines.

File: packages/block-editor/src/hooks/layout.js
Comment (jsnajdr) [inline packages/block-editor/src/hooks/layout.js:486]: There's one more destructuring instance to fix in `BlockWithChildLayoutStyles`.

**Review by jsnajdr — APPROVED** [review]
Thanks for addressing the feedback :+1:

File: packages/block-editor/src/hooks/layout.js
Comment (tellthemachines) [reply packages/block-editor/src/hooks/layout.js:474]: Oh that sounds like a bug, yes. The child layout CSS shouldn't be generated at all if `disableLayoutStyles` is true.

**Review by tellthemachines — APPROVED** [review]
Changes LGTM and nothing seems broken when testing different blocks with layout ✅ 

> As a followup, would it make sense to merge the withLayoutStyles and withChildLayoutStyles HOCs?

I don't think so because their logic has very little in common. Pretty much the only common part is the use of `setStyleOverride` in an effect at the end of each of the functions.

**Comment by Mamaduka** [comment]
Thanks for testing, @tellthemachines 🙇‍♂️

**Comment by tellthemachines** [comment]
Looks like this change introduced a weird bug in the ChildLayoutControls where [the effect](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/child-layout-control/index.js#L43) that's only supposed to run once on mount is actually running again after a change is made. From my testing it only happens after the first change, _if_ the block has no child layout set. It's only noticeable in the UI when trying to change the child layout to Fixed, because the input doesn't appear below.

I'm not sure what in this commit could have caused this, but it's definitely here that the bug was born 😅 

Edit: here's a screenshot of the weird intermediate state with "Fixed" selected and no input:

Screenshot 2023-11-29 at 3 42 42 pm

**Comment by Mamaduka** [comment]
@tellthemachines, can you share the reproduction steps? I'll look into this today.

Thank you 🙇

**Comment by tellthemachines** [comment]
@Mamaduka this is reproducible in 17.1.x and in trunk. 

1. Add a Row block with some children to a post or template;
2. Select one of the children, and in the sidebar, under Dimensions (in the Styles tab if the tabs appear), try changing width to "Fixed".
3. See that there's no input appearing under the width controls.

You can also compare against an install of WP 6.4 without Gutenberg as that works as expected.

**Comment by Mamaduka** [comment]
This is a regression due to the `BlockListBlock` remount. Here's what's happening:

1. The `shouldRenderChildLayoutStyles` value is initially `false` and HOC renders the original  `BlockListBlock`.
2. The effect in `ChildLayoutControl` does nothing since the condition isn't met.
3. Changing the `selfStretch` value also changes the `shouldRenderChildLayoutStyles` to `true`, rendering the `BlockWithChildLayoutStyles` component.
4. The effect meets the condition in freshly mounted `ChildLayoutControl`, changing `selfStretch` to `fit` and hiding the unit control.

Oddly, the toggle option group selected value stays the same, while the rest of the component behaves "as expected." It's probably worth investigating separately.

I don't have a perfect solution besides reverting the changes, but I'll keep looking.

@jsnajdr, we might want to re-evaluate [similar perf optimizations](https://github.com/WordPress/gutenberg/pull/55449/) and apply same fix.

**Comment by jsnajdr** [comment]
I'm not sure if I'll be able to look at this today, but what has been described so far sounds like some React props and data changes are not propagated correctly. A typical React bug. For example, the effect in `ChildLayoutControl` runs only on mount, and ESLint is warning about missing dependencies. Maybe the effect should work differently? If I understand it correctly, it resets the `selfStretch` back to `fit` when it has an "invalid" `fixed` value, without a `flexSize` specified. Should it really run only on mount, or should it watch the `selfStretch` and `flexSize` changes during the entire lifetime of the component?

**Comment by Mamaduka** [comment]
Right, assuming that the component will only mount once isn't guaranteed in React, and we should look into a fix for the `ChildLayoutControl` component.

I think we should also fix the `BlockListBlock` being remounted; I don't have the numbers, but remounting large branches for blocks shouldn't be good for performance. What do you think?

**Comment by tellthemachines** [comment]
> If I understand it correctly, it resets the `selfStretch` back to `fit` when it has an "invalid" `fixed` value, without a `flexSize` specified. Should it really run only on mount, or should it watch the `selfStretch` and `flexSize` changes during the entire lifetime of the component?

The effect should only run on mount, because otherwise it will be impossible to set the value to "Fixed" 😅 the problem being that, the way the controls are designed, we need to first set the toggle to "Fixed" in order to display the input where we can then add a fixed value.

I guess this wouldn't be a problem if the controls were designed differently, and given there has been talk of that for a while (e.g. in #46128) this effect was seen as an acceptable hack to make things work in the meantime.

**Comment by tellthemachines** [comment]
Thanks for looking into this folks!

#56660 partially fixes this issue (the input now displays correctly) but with the toggle not resetting from "Fixed" to "Fit" when there is no value assigned remains. Looking into that now.

Update: the toggle not resetting seems to be caused by the value [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/toggle-group-control/toggle-group-control/as-radio-group.tsx#L53) always being undefined (the actual intended value of the control is being set as `defaultValue` instead). Cc @ciampo who most recently edited that code.

Second update: the above issue happens due to the effect in `useComputeControlledOrUncontrolledValue` running after [the code beneath it](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/toggle-group-control/toggle-group-control/utils.ts#L41), so that the component returns its original value as it the effect had never run. @ciampo I'm not sure what the intent of the effect was in there; could we perhaps do without it?

Third update: #56660 is now merged and I've created #56667 for the toggle issue.

**Comment by jsnajdr** [comment]
> The effect should only run on mount, because otherwise it will be impossible to set the value to "Fixed" 😅

Yeah, now that I got to have a detailed look at this issue, I understand that better. 🙂 

> this effect was seen as an acceptable hack to make things work in the meantime.

One thing that's bad about the effect is that it pollutes the undo stack with changes of the `style.layout.selfStretch` attribute from `fit` to `fixed` and back.

Ideally, the `onChange` callback, which modifies the block attribute and triggers the entire machinery of modifying a post (undo, syncing the REST entity...), would be called only after the entire width-changing operation is finished with a valid result. I.e., `selfStretch` is `fixed` and `flexSize` is a valid dimension. Not on every click with intermediate invalid values. In the meantime, `ChildLayoutControl` should keep the work-in-progress values in local state.

**Comment by tellthemachines** [comment]
> One thing that's bad about the effect is that it pollutes the undo stack with changes of the `style.layout.selfStretch` attribute from `fit` to `fixed` and back.

We could perhaps use `__unstableMarkNextChangeAsNotPersistent` as a workaround for that.

**Comment by andrewserong** [comment]
I think this PR might have caused a regression in 6.5 (#60569) — it looks like this PR changed the logic so that if `disableLayoutStyles` is set, then the output of semantic classnames is skipped. It looks like that likely wasn't intentional, but just thought I'd double check in case it was.

**Pull request merged**
